### PR TITLE
Compress large json cache data (i.e. undo what couchbase guys did)

### PR DESCRIPTION
**Pull request by raymondng**
Using default gzip compression for now as the support is already in the memcached client. Will collect data and go from there. @andrewconner @stkem @gmethvin @ymatsuda

### CODE REVIEW

File: kifi-backend/modules/common/test/com/keepit/common/cache/FortyTwoCacheTest.scala [hunk 1]
```diff
@@ -78,17 +78,17 @@ class FortyTwoCacheTest extends Specification with DeprecatedTestInjector {
                 cache.set(TestJsonCacheKey("missing_value"), None)
                 cache.getOrElseOpt(TestJsonCacheKey("missing_value"))(Some(TestJsonCacheData("Error!", -42))) === None
             }
-            "throw exception & invalidate cache when size exceeded" in {
```
Comment (eishay) [inline kifi-backend/modules/common/test/com/keepit/common/cache/FortyTwoCacheTest.scala:81]: why are the tests commented out?

File: kifi-backend/modules/common/app/com/keepit/common/cache/MemcachedPlugin.scala [hunk 2]
```diff
@@ -35,11 +35,44 @@ class MemcachedSlf4JLogger(name: String) extends AbstractLogger(name) {
 @Singleton
 class MemcachedPlugin @Inject() (client: MemcachedClient) extends CachePlugin {
 
+  val compressThreshold:Int = 400000 // TODO: make configurable
+  val compressMethod:String = "gzip"
```
Comment (stkem) [inline kifi-backend/modules/common/app/com/keepit/common/cache/MemcachedPlugin.scala:39]: Didn't we determine that gzip was too slow? Or is this doing something different from @andrewconner s benchmark?

File: kifi-backend/modules/common/test/com/keepit/common/cache/FortyTwoCacheTest.scala (hunk 1 above)
Comment (raymondng) [inline kifi-backend/modules/common/test/com/keepit/common/cache/FortyTwoCacheTest.scala:81]: The check in FortyTwoCache has been commented out -- the real check is now done in MemcachedPlugin. 

The unit tests currently do not run with memcached enabled.

I was able to manually test this locally using a local memcached client, and I'd like to automate this (integration test with local memcached) going forward, assuming everyone is ok with it.

File: kifi-backend/modules/common/app/com/keepit/common/cache/MemcachedPlugin.scala [hunk 3]
```diff
@@ -35,11 +35,44 @@ class MemcachedSlf4JLogger(name: String) extends AbstractLogger(name) {
 @Singleton
 class MemcachedPlugin @Inject() (client: MemcachedClient) extends CachePlugin {
 
+  val compressThreshold:Int = 400000 // TODO: make configurable
```
Comment (eishay) [inline kifi-backend/modules/common/app/com/keepit/common/cache/MemcachedPlugin.scala:38]: very nice. I'm not sure you need to make it configurable. Its not something I expect to change from service to service or different environments.

File: kifi-backend/modules/common/app/com/keepit/common/cache/MemcachedPlugin.scala [hunk 4]
```diff
@@ -35,11 +35,44 @@ class MemcachedSlf4JLogger(name: String) extends AbstractLogger(name) {
...
+            val ts = System.currentTimeMillis
+            val compressed = compressMethod match {
+              case "gzip" => compress(b)
+              case _ => throw new UnsupportedOperationException(s"Compression method $compressMethod not supported")
+            }
+            val lapsed = System.currentTimeMillis - ts
+            if (compressed.length < b.length) {
+              logger.info(s"Compressed ${o.getClass.getName}: ${b.length} => ${compressed.length} in $lapsed ms")
+              b = compressed
+              flags = 2 // COMPRESSED -- @see SerializingTranscoder
```
Comment (eishay) [inline kifi-backend/modules/common/app/com/keepit/common/cache/MemcachedPlugin.scala:62]: what is `flags = 1` ?

File: kifi-backend/modules/common/app/com/keepit/common/cache/MemcachedPlugin.scala [hunk 5]
```diff
@@ -35,11 +35,44 @@ class MemcachedSlf4JLogger(name: String) extends AbstractLogger(name) {
...
+            } else {
+              logger.warn(s"Compression INCREASED size of ${o.getClass.getName}: ${b.length} => ${compressed.length} in $lapsed ms")
+            }
+          }
+          if (b.length > maxThreshold) {
+            throw new CacheSizeLimitExceededException(s"object $o size=${b.length} is greater than memcached max threshold=$maxThreshold")
+          }
+          new CachedData(flags, b, getMaxSize)
```
Comment (eishay) [inline kifi-backend/modules/common/app/com/keepit/common/cache/MemcachedPlugin.scala:70]: cool

File: kifi-backend/modules/common/app/com/keepit/common/cache/MemcachedPlugin.scala (hunk 2 above)
Comment (raymondng) [inline kifi-backend/modules/common/app/com/keepit/common/cache/MemcachedPlugin.scala:39]: That's the default mechanism built into the client -- it's just a start. I don't doubt @andrewconner's benchmarks and I'd also like to gather some data in prod. 
We will certainly explore alternatives if this is not acceptable.

Comment (raymondng) [inline kifi-backend/modules/common/app/com/keepit/common/cache/MemcachedPlugin.scala:39]: FWIW the compressMethod is meant to be made configurable as well.

File: kifi-backend/modules/common/app/com/keepit/common/cache/MemcachedPlugin.scala (hunk 4 above)
Comment (raymondng) [inline kifi-backend/modules/common/app/com/keepit/common/cache/MemcachedPlugin.scala:62]: serialized

**Comment by andrewconner** [comment]
Looks good to me, nice.

**Pull request merged**